File: app/src/main/java/space/pal/sig/view/splash/SplashViewModel.java
package space.pal.sig.view.splash;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkInfo;
import androidx.work.WorkManager;

import space.pal.sig.util.SharedPreferencesUtil;

/**
 * SpaceNow
 * Created by Catalin on 7/22/2020
 **/
public class SplashViewModel extends AndroidViewModel {

    private static final String IS_FIRST_TIME = "Is.First.Time";
    private static final String TAG_OUTPUT = "Worker.Tag.Output";
    private final WorkManager manager;
    private final SharedPreferencesUtil sharedPreferencesUtil;
    private final MediatorLiveData<Boolean> safeForStart = new MediatorLiveData<>();

    public SplashViewModel(@NonNull Application application,
                           SharedPreferencesUtil sharedPreferencesUtil) {
        super(application);
        this.manager = WorkManager.getInstance(application);
        this.sharedPreferencesUtil = sharedPreferencesUtil;
        boolean isFirstTime = isFirstTime();
        if (isFirstTime) {
            setupWork();
        }
        this.safeForStart.setValue(!isFirstTime);
    }

    public LiveData<Boolean> isSafeForStart() {
        return safeForStart;
    }

    private void setupWork() {
        safeForStart.addSource(manager.getWorkInfosByTagLiveData(TAG_OUTPUT),
                workInfoList -> {
                    WorkInfo workInfo = workInfoList.get(0);
                    boolean finished = workInfo.getState().isFinished();
                    if (finished) {
                        sharedPreferencesUtil.storeBoolean(IS_FIRST_TIME, false);
                    }
                    safeForStart.setValue(finished);
                });
        OneTimeWorkRequest worker = new OneTimeWorkRequest
                .Builder(DataLoadWorker.class)
                .addTag(TAG_OUTPUT)
                .build();
        manager.enqueue(worker);
    }

    private boolean isFirstTime() {
        return sharedPreferencesUtil.getBoolean(IS_FIRST_TIME, true);
    }
}

File: app/src/main/java/space/pal/sig/view/splash/SplashViewModel.kt
package space.pal.sig.view.splash

import android.app.Application
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import space.pal.sig.service.DataSyncWorker
import space.pal.sig.service.ScheduleManager
import space.pal.sig.util.SharedPreferencesUtil
import space.pal.sig.view.BaseViewModel

/**
SpaceNow
Created by Catalin on 12/13/2020
 **/
class SplashViewModel(
        application: Application,
        sharedPreferencesUtil: SharedPreferencesUtil
) : BaseViewModel(application) {

    companion object {
        const val IS_FIRST_TIME = "SpaceNow.IsFirstTime"
    }

    private val safeForStart = MediatorLiveData<Boolean>()
    private val syncResult = ScheduleManager
            .launchOneTimeSync(application, DataSyncWorker.DATA_SYNC_WORKER_TAG)

    init {
        val isFirstTime = sharedPreferencesUtil.get(IS_FIRST_TIME, true)
        if (isFirstTime) {
            safeForStart.addSource(syncResult) { workInfoList ->
                val finished = workInfoList[0].state.isFinished
                safeForStart.value = finished
                sharedPreferencesUtil.save(IS_FIRST_TIME, false)
            }
        } else {
            safeForStart.value = true
        }
    }

    fun isSafeForStart(): LiveData<Boolean> {
        return safeForStart
    }

}